### PR TITLE
[API Compat] Add rule to validate an abstract member is not added

### PR DESCRIPTION
**Pull request by safern**
Contributes towards: https://github.com/dotnet/sdk/issues/17928

FYI: @terrajobst @utpilla

### CODE REVIEW

File: src/Compatibility/Microsoft.DotNet.ApiCompatibility/Rules/CannotAddAbstractMember.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        private void RunOnMemberMapperAction(ISymbol left, ISymbol right, ITypeSymbol leftContainingType, ITypeSymbol rightContainingType, string leftName, string rightName, IList<CompatDifference> differences)
+        {
+            if (left == null && right != null && right.IsAbstract)
+            {
+                // We need to make sure left containing type is not sealed, as unsealing a type is not a breaking change.
+                // So if in this version of left and right, right is unsealing the type, abstract members can be added.
+                // checking for member additions on interfaces is checked on its own rule.
+                if (leftContainingType.TypeKind != TypeKind.Interface && !leftContainingType.IsSealed && HasVisibleConstructor(leftContainingType as INamedTypeSymbol))
+                {
+                    differences.Add(new CompatDifference(DiagnosticIds.CannotAddAbstractMember, $"Cannot add abstract member '{right.ToDisplayString()}' to {rightName} because does not exist on {leftName}", DifferenceType.Added, right));
```
Comment (safern) [inline src/Compatibility/Microsoft.DotNet.ApiCompatibility/Rules/CannotAddAbstractMember.cs:41]: Need to add this to the resource strings.

File: src/Compatibility/Microsoft.DotNet.ApiCompatibility/Rules/CannotAddAbstractMember.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
```
Comment (Anipik) [inline src/Compatibility/Microsoft.DotNet.ApiCompatibility/Rules/CannotAddAbstractMember.cs:3]: nit: extra ```//```

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/CannotAddAbstractMemberTests.cs [hunk 3]
```diff
@@ -0,0 +1,260 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
```
Comment (Anipik) [inline src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/CannotAddAbstractMemberTests.cs:3]: nit extra ```//```

File: src/Compatibility/Microsoft.DotNet.ApiCompatibility/DiagnosticIds.cs [hunk 4]
```diff
@@ -10,5 +10,6 @@ public static class DiagnosticIds
     {
         public const string TypeMustExist = "CP0001";
         public const string MemberMustExist = "CP0002";
+        public const string CannotAddAbstractMember = "CP0004";
```
Comment (Anipik) [inline src/Compatibility/Microsoft.DotNet.ApiCompatibility/DiagnosticIds.cs:13]: we might need to change it CP0005

**Review by Anipik — APPROVED** [review]
LGTM After adding the string change.

File: src/Compatibility/Microsoft.DotNet.ApiCompatibility/DiagnosticIds.cs (hunk 4 above)
Comment (safern) [reply src/Compatibility/Microsoft.DotNet.ApiCompatibility/DiagnosticIds.cs:13]: Yeah, I'll do that. Thanks!

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/CannotAddAbstractMemberTests.cs [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+  }
+}
+",
+                @"
+namespace CompatTests
+{
+  public class First
+  {
+    public abstract void FirstAbstract() { }
+    public abstract string SecondAbstract() => throw null;
```
Comment (ericstj) [inline src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/CannotAddAbstractMemberTests.cs:267]: Are these meant to be syntactically correct?  https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANmgExAGoAfAOwEMBbGAZwAdKwYACAYQmqYwBV6MdALAAoAN6jWrAAIBmGQCZWAMWxQ6GSawkipUua0rANUZhhkoVajQEFjGU2AwAKAJTbWAXy375Rk2YyAIwADKwAyjCQ5AR2AU5urAC8AHysGAAWUBAA7qzkAK64uADcWt4inkA

Comment (safern) [reply src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/CannotAddAbstractMemberTests.cs:267]: Good catch. They shouldn't affect the tests, but we I should fix them.

**Pull request merged**